src/model_signing/_serialization/incremental.py

```diff
@@ -0,0 +1,247 @@
+"""
+
+from collections.abc import Callable, Iterable
+import concurrent.futures
+import itertools
+import os
+import pathlib
+from typing import Optional
+
+from typing_extensions import override
+
+from model_signing import manifest
+from model_signing._hashing import io
+from model_signing._serialization import serialization
+
+
+class IncrementalSerializer(serialization.Serializer):
+    """Model serializer that only re-hashes changed files.
+
+    This serializer compares the current model state against an existing
+    manifest (from a previous signature) and only re-hashes files that:
+    - Are new (not in the existing manifest)
+    - Have changed size (likely modified)
+    - Are explicitly requested via files_to_hash parameter
+
+    Files that exist in both the current model and the existing manifest
+    with matching sizes will have their digests reused from the existing
+    manifest without re-hashing.
+
+    This provides significant performance improvements for large models where
+    only a small number of files change between signings (e.g., updating
+    documentation in a 200GB model).
+    """
+
+    def __init__(
+        self,
+        file_hasher_factory: Callable[[pathlib.Path], io.FileHasher],
+        existing_manifest: manifest.Manifest,
+        *,
+        max_workers: Optional[int] = None,
+        allow_symlinks: bool = False,
+        ignore_paths: Iterable[pathlib.Path] = frozenset(),
+    ):
+        """Initializes an incremental serializer.
+
+        Args:
+            file_hasher_factory: A callable to build the hash engine used to
+              hash individual files.
+            existing_manifest: The manifest from a previous signature. Digests
+              from this manifest will be reused for unchanged files.
+            max_workers: Maximum number of workers to use in parallel. Default
+              is to defer to the `concurrent.futures` library.
+            allow_symlinks: Controls whether symbolic links are included. If a
+              symlink is present but the flag is `False` (default) the
+              serialization would raise an error.
+            ignore_paths: The paths of files to ignore.
+        """
+        self._hasher_factory = file_hasher_factory
+        self._existing_manifest = existing_manifest
+        self._max_workers = max_workers
+        self._allow_symlinks = allow_symlinks
+        self._ignore_paths = ignore_paths
+
+        # Build lookup dictionary: file path -> manifest item
+        self._existing_items = {}
+        for item in existing_manifest._item_to_digest:
+            # item is a _File or _Shard key; we only support files for now
+            if isinstance(item, manifest._File):
+                self._existing_items[item.path] = item
+
+        # Precompute serialization description
+        hasher = file_hasher_factory(pathlib.Path())
+        self._serialization_description = manifest._FileSerialization(
```

Comment:
This would be the same as `manifest.serialization_type()`, right?

Reply:
Oh, that's a nicer improvement!

Reply:
I think this should still just copy from the existing manifest (similar to a C++ copy-constructor)